### PR TITLE
Auto-generate shell completions for solana-cli (issue #8879 and #14005)

### LINKED ISSUE #8879: No tab completion in CLI
#### Problem

No tab completion in CLI, forcing me to ask for `--help` more than I'd like.

#### Proposed Solution

https://clap.rs/2016/10/25/an-update/#completionscriptgenerationinclap

### PR DESCRIPTION
**Pull request by theonekeyg**
Implement `completion` SubCommand for solana-cli, which outputs
completion script to stdout and exits the process. The script generation
handled completely by clap.

In order to implement the generation, one minor design change was
necessary regarding the creation of clap `App`.

**Previously**: One part of App initialization was in the `app` function,
and some other arguments and subcommands were added later directly in
the `main` function.

**Now**: The whole construction of App was moved to `get_clap_app` function.

**P.S.** I wasn't sure if constructing App separately had visual importance,
so both constructing parts are still separate in `base_clap_app` and
`finalize_clap_app` functions. But they sure could be one single function.

#### Problem

Absence of shell completions in solana-cli. (#8879 and #14005)

#### Summary of Changes

* Added `completion` subcommand for solana-cli to generate completion scripts.
* Tiny rework of App initialization.

Fixes #8879 and #14005

### CODE REVIEW

**Review by CriesofCarrots — COMMENTED** [review]
Thanks! I've been wanting this for a while.

I actually don't think there's a strong reason the `base_` and `finalize_` parts of the App have to remain separate anymore. Wdyt about adding a separate module to hold the app-construction function?

@t-nelson , can you take a look at this as well?

File: cli/src/cli.rs [hunk 1]
```diff
@@ -949,6 +953,37 @@ pub fn parse_command(
...
+                        crate_name!(), crate_description!(),
+                        solana_version::version!()
+                    ).gen_completions_to("solana", Shell::PowerShell, &mut stdout()),
+                Some("elvish") =>
+                    get_clap_app(
+                        crate_name!(), crate_description!(),
+                        solana_version::version!()
+                    ).gen_completions_to("solana", Shell::Elvish, &mut stdout()),
+                _ => unreachable!() // This is safe, since we assign default_value
+            }
```
Comment (t-nelson) [inline cli/src/cli.rs:984]: nit: could dereplicode this block by having the `match` only select the `Shell` variant, then make a single call to `gen_completions_to`

**Review by t-nelson — APPROVED** [review]
Looks great!  I've been wanting these for awhile too

**Comment by t-nelson** [comment]
> I actually don't think there's a strong reason the base_ and finalize_ parts of the App have to remain separate anymore. Wdyt about adding a separate module to hold the app-construction function?

Agreed.  Let's tie these up while they're getting shuffled around anyway

**Review by CriesofCarrots — APPROVED** [review]
Looks great, thank you so much!

**Pull request merged**